### ISSUE TITLE
Register custom compression types via entrypoint mechanism

### ISSUE THREAD

**Issue by cjackal** [fsspec/filesystem_spec#1187]
While new `fsspec`-compatible filesystem can easily be registered via `fsspec.specs` entry points, there is no counterpart for compression types.

As we already have the `fsspec.compression.register_compression` public API for custom compression types, it wouldn't be too much burden (and give better user experience as a universal filesystem interface) to implement entrypoint support for custom compression types.

THOUGHTS.

`register_compression` takes two names(in addition to the (de-)compressor callable): algorithm name(e.g. gzip) and file extension(e.g. gz). Receiving both names via entrypoint is not impossible (simple parsing in mind) but may not be desirable from aesthetic viewpoint. In any case registration with whatever the preferred name can still be beneficial, as users will be able to specify the compression method via simple string argument, e.g. `fsspec.open(..., compression=compression)`, without import-and-call `register_compression` manually.

(BTW this feature request comes after my reading a bunch of zlib-compressed csv files. I just found that zlib is not supported in fsspec and can be implemented in almost identical manner as how snappy is implemented.)

**Comment by milesgranger** [comment]
> no file-like API

I assume you're talking about things like `cramjam.snappy.compress(...)` to take a file-like object? As [cramjam.Buffer](https://docs.rs/cramjam/latest/cramjam/io/struct.RustyBuffer.html) is file-like. Wouldn't be a bad idea to add support for general file-like objs though. :+1:

**Comment by martindurant** [comment]
Yes of course - I mean the other way around, that cannot stream decompress from a file in the style of something like gzip.GZipFile. For example, zlib/gzip/inflate in crate flate2 seem to operate on [buffered]read streamz. These streams could in principle be python file-like objects, but calling back into fsspec seems like a bad idea, since it may need to wait on an event loop in another thread.

OTOH, https://github.com/martindurant/rfsspec shows that concurrency purely in rust and in-thread is possible, so we could build the streams that way, with transparent decompression. Since making that POC last week, I have been thinking about how much of fsspec could/should be pushed into rust, and you @milesgranger  might just be the person with the best feel for this.
